src/totp/totp.ts

```diff
@@ -0,0 +1,122 @@
+};
+
+export class TOTP {
+  /**
+   * Generates a Time-based One-Time Password (TOTP).
+   * @param {Uint8Array} rawKey - The secret key for TOTP.
+   * @param {TOTPOptions} options - Optional parameters for TOTP.
+   * @returns {string} The generated TOTP.
+   */
+  static generate(rawKey: Uint8Array, options?: TOTPOptions): string {
+    const _options = this.withDefaultOptions(options);
+
+    const counter = this.getCounter(_options.timestamp, _options.period);
+    const otp = generateHOTP(rawKey, counter, _options);
+    return otp;
+  }
+
+  /**
+   * Validates a Time-based One-Time Password (TOTP).
+   * @param {Uint8Array} rawKey - The secret key for TOTP.
+   * @param {string} passcode - The passcode to validate.
+   * @param {TOTPValidateOptions} options - Optional parameters for TOTP validation.
+   * @returns {boolean} A boolean indicating whether the passcode is valid.
+   */
+  static validate(rawKey: Uint8Array, passcode: string, options?: TOTPValidateOptions): boolean {
+    const _options = this.withDefaultValidateOptions(options);
+    passcode = passcode.trim();
+    if (passcode.length != _options.digits) {
+      return false;
+    }
+
+    const counter = this.getCounter(_options.timestamp, _options.period);
+
+    const counters = [counter];
+    for (let i = 1; i <= _options.skew; i++) {
+      counters.push(counter + i);
+      counters.push(counter - i);
+    }
+
+    for (let c of counters) {
+      if (passcode === generateHOTP(rawKey, c, _options)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private static getCounter(timestamp: number, period: number): number {
+    const epochSeconds = Math.floor(timestamp / 1000);
+    const counter = Math.floor(epochSeconds / period);
+    return counter;
+  }
+
+  private static withDefaultOptions(options?: TOTPOptions): Required<TOTPOptions> {
+    return {
+      digits: 2,
+      algorithm: 'SHA-1',
+      period: 30,
+      timestamp: Date.now(),
+      ...options,
+    };
+  }
+
+  private static withDefaultValidateOptions(options?: TOTPValidateOptions): Required<TOTPValidateOptions> {
+    return { skew: 1, ...this.withDefaultOptions(options) };
+  }
+}
+
+function generateHOTP(rawKey: Uint8Array, counter: number, options: Required<TOTPOptions>): string {
+  const keyArray = Array.from(rawKey);
+  const timeHex = dec2hex(counter).padStart(16, '0');
+  const hmac = calcHMAC(keyArray, timeHex, options.algorithm);
+
+  const signatureHex = hmac.digestHex();
+
+  const offset = hex2dec(signatureHex.slice(-1)) * 2;
+  const masked = hex2dec(signatureHex.slice(offset, offset + 8)) & 0x7fffffff;
+  const otp = masked.toString().slice(-options.digits);
+  return otp;
+}
+
+function calcHMAC(keyArray: number[], timeHex: string, algorithm: TOTPAlgorithm) {
+  switch (algorithm) {
+    case 'SHA-1':
+      // Because SHA1HMAC is missing in @bsv/sdk, we had to implement it (using SHA1 from @bsv/sdk)
+      // TODO: Consider adding SHA1HMAC to @bsv/sdk

```

Comment:
It would be great to create an issue in "Discussion" type:
<img width="1403" alt="image" src="https://github.com/bitcoin-sv/spv-wallet-js-client/assets/48433067/c9407a78-80f1-421f-914e-48294942943d">
for them to consider and add this feature to the library :D 

// I don't like hanging TODOs :D

Reply:
What I was planned is to show `ts-sdk` team the whole `totp` directory and discuss what to move to their package.
However, I'd prefer to do it after our internal rewiews & tests - so basically, after merging-  because I would not like the discussion to block this PR.